x-pack/elastic-agent/pkg/core/plugin/app/app.go

```diff
@@ -206,16 +212,35 @@ func (a *Application) waitProc(proc *os.Process) <-chan *os.ProcessState {
 	return resChan
 }
 
-func (a *Application) reportCrash(ctx context.Context) {
-	a.monitor.Cleanup(a.name, a.pipelineID)
+func (a *Application) setStateFromProto(pstatus proto.StateObserved_Status, msg string) {
+	var status state.Status
+	switch pstatus {
+	case proto.StateObserved_STARTING:
+		status = state.Starting
+	case proto.StateObserved_CONFIGURING:
+		status = state.Configuring
+	case proto.StateObserved_HEALTHY:
+		status = state.Running
+	case proto.StateObserved_DEGRADED:
+		status = state.Degraded
+	case proto.StateObserved_FAILED:
+		status = state.Failed
+	case proto.StateObserved_STOPPING:
+		status = state.Stopping
+	}
+	a.setState(status, msg)
+}
 
-	// TODO: reporting crash
-	if a.failureReporter != nil {
-		crashError := errors.New(
-			fmt.Sprintf("application '%s' crashed", a.id),
-			errors.TypeApplicationCrash,
-			errors.M(errors.MetaKeyAppName, a.name),
-			errors.M(errors.MetaKeyAppName, a.id))
-		a.failureReporter(ctx, a.name, crashError)
+func (a *Application) setState(status state.Status, msg string) {
+	if a.state.Status != status || a.state.Message != msg {
+		a.state.Status = status
+		a.state.Message = msg
+		if a.reporter != nil {
+			go a.reporter.OnStateChange(a.id, a.name, a.state)
```

Comment:
what in case of fast change of state changes as goroutines dont have guaranteed order of start this can be tricky

Reply:
Because the old `reportCrash` used a go routine, I kept this the same way. But your point is very valid. I think we should change `Reporter` to keep a queue of records to send and send then in a dedicated go routine that way we can ensure the order is consistent.

I can do that in a follow up branch.